### PR TITLE
SearchAutocomplete

### LINKED ISSUE #2084: SearchAutocomplete
SearchAutocomplete is essentially a ComboBox without the dropdown button. It is commonly seen as a search input or dynamic search. Because SearchAutocomplete will implement the Collections API (both static and dynamic), adding it as a variant to the existing SearchField would cause unnecessary bloat for those with simple usecases.

SearchAutocomplete consists of a SearchField and a Popover Listbox.

dynamic-search_anatomy@2x_2EO0r9PRrwKd3Ye60veyQ3_1611703894000

### Example
``` jsx
<SearchAutocomplete label="Search">
  <Item>Foo</Item>
  <Item>Bar</Item>
</SearchAutocomplete>

<SearchAutocomplete label="Search">
  {(item) => <Item>{item.name}</Item>}
</SearchAutocomplete>
```

### API
``` tsx

interface SearchAutocompleteProps<T> extends CollectionBase<T>, SearchFieldProps {
  menuTrigger: 'focus' | 'input',
  isQuiet: boolean,
  direction: 'bottom' | 'top',
  shouldFlip?: boolean,
  loadingState: LoadingState,
  onLoadMore?: () => void,
  onSubmit?: (value: string, key: Key | null) => void,
  disabledKeys?: Set<Key>
}
```

### Implementation
To start with we want to only use the autocomplete with the `SearchField`, therefore it makes sense to add it to the `@react-spectrum/searchfield`  package. In addition, much of the internals for autocomplete will come from the combobox package so we will need to export those needed pieces. In the future, we hope to add autocomplete to other components beyond SearchField. 

We will be creating a generic autocomplete component, however it will not be public.  This internal generic autocomplete component will need to accept an input element (SearchField) and because of our react spectrum mobile solution, a “fake” input that is actually a button that opens the Tray similar to ComboBox.

For the react aria portion, there should be a `useSearchAutocomplete` hook that combines `useComboBox` and `useSearchField`. 

Because a search field is more oriented around the search text and custom values rather than the keys of the items,  `selectedKey`/`onSelectionChange` will not be supported. For cases where users want to know which key was specifically selected, the `onSubmit` callback will receive the key in addition to the search text. Additionally, `SearchAutocomplete` will always allow custom values.

#### Accessibility open questions
- The role of the SearchField - role="searchbox" or role="combobox"? aria-autocomplete is allowed on both roles, but we should research/test whether it is supported and works well across screen readers.

## 🧢 Your Company/Team
RSP

### PR DESCRIPTION
**Pull request by reidbarber**
Closes #2084

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Check SearchWithin sections of Docs and Storybook, along with tests and chromatic stories.

### CODE REVIEW

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+
+```jsx import
+import {SearchAutocomplete, Section, Item} from '@react-spectrum/searchfield';
+import {useFilter} from '@react-aria/i18n';
+```
+
+---
+category: Pickers
+keywords: [search field, input]
+---
```
Comment (reidbarber) [inline packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:25]: What should `after_version` be here?

**Comment by reidbarber** [comment]
Still working on:
- [ ] Tests
- [ ] Chromatic stories
- [ ] Selection from listbox should trigger onSubmit with new value and key
- [ ] Fix style for quiet on mobile when input has value
- [ ] Deduplicate focus/blur events
- [ ] 'Esc' key while typing on mobile closes and re-opens tray

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx (hunk 1 above)
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:26]: fixed to `after_version: 3.0.0`

**Comment by LFDanLu** [comment]
@reidbarber Mind getting rid of the yarn error file 😄. Also looks like you have some package json conflicts

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+    shouldCloseOnBlur: false,
+    allowsCustomValue: true
+  });
+
+  let searchFieldState = useSearchFieldState({
+    ...props,
+    onSubmit: (value) => onSubmit(value, null)
+  });
+
+  let state = {...comboBoxState, ...searchFieldState};
```
Comment (LFDanLu) [inline packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx:77]: urgh, this feels iffy since there is a state tracker for `inputValue` and `value`. Why was this needed here but not in SearchAutocomplete?

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+            styles,
+            'spectrum-InputGroup',
+            {
+              'spectrum-InputGroup--quiet': isQuiet,
+              'is-disabled': isDisabled,
+              'spectrum-InputGroup--invalid': validationState === 'invalid',
+              'is-hovered': isHovered
+            },
+            classNames(
+              comboboxStyles,
```
Comment (LFDanLu) [inline packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx:238]: might be better to just duplicate these styles in the searchfield package or move them into spectrum css so you don't need to import the `react-spectrum/combobox` package

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+      timeout.current = null;
+    }
+
+    lastInputValue.current = inputValue;
+  }, [loadingState, inputValue, showLoading]);
+
+  let onKeyDown = (e) => {
+    // Close virtual keyboard if user hits Enter w/o any focused options
+    if (e.key === 'Enter' && state.selectionManager.focusedKey == null) {
+      popoverRef.current.focus();
```
Comment (LFDanLu) [inline packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx:449]: Might need to fire `onSubmit` manually here as well

**Review by LFDanLu — COMMENTED** [review]
Feel free to address these in a followup PR

File: packages/@react-spectrum/searchfield/package.json [hunk 5]
```diff
@@ -31,15 +31,39 @@
     "url": "https://github.com/adobe/react-spectrum"
   },
   "dependencies": {
+    "@adobe/react-spectrum": "^3.12.0",
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/package.json:34]: Good catch, probably from a wrong auto-import from the IDE

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+      defaultItems={options}
+      placeholder="Aerospace, Mechanical, etc.">
+      {item => <Item>{item.name}</Item>}
+    </SearchAutocomplete>
+  );
+}
+```
+
+## Labeling
+SearchAutocomplete can be labeled using the `label` prop. If the SearchAutocomplete is a required field, the `isRequired` and `necessityIndicator` props can be used to show a required state.
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:123]: Left this in because we should probably support whatever SearchField does, and it currently supports these. I'm open to removing these though; you definitely have a good point. I think we had a similar discussion for SearchWithin and didn't end up removing it.

File: packages/@react-spectrum/searchfield/test/SearchAutocomplete.test.js [hunk 7]
```diff
@@ -0,0 +1,2664 @@
...
+  afterEach(() => {
+    jest.clearAllMocks();
+    act(() => jest.runAllTimers());
+  });
+
+  afterAll(function () {
+    jest.restoreAllMocks();
+  });
+
+  it('renders correctly', function () {
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/test/SearchAutocomplete.test.js:159]: Yes, they mostly test the same thing but use typing to trigger instead of combobox's button trigger.

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx (hunk 3 above)
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx:238]: Good idea, I'll do that.

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx (hunk 2 above)
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx:77]: Good find, this should be updated to look like SearchAutocomplete. I'll fix that.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+governing permissions and limitations under the License. -->
+
+import {Layout} from '@react-spectrum/docs';
+export default Layout;
+
+import docs from 'docs:@react-aria/searchfield';
+import collectionsDocs from 'docs:@react-types/shared/src/collections.d.ts';
+import {FunctionAPI, HeaderInfo, InterfaceType, TypeContext, TypeLink} from '@react-spectrum/docs';
+import i18nDocs from 'docs:@react-aria/i18n';
+import listboxDocs from 'docs:@react-aria/listbox';
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Good catch, fixed.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+  />
+
+## API
+
+<FunctionAPI function={docs.exports.useSearchAutocomplete} links={docs.links} />
+
+## Features
+
+Autocomplete for search fields can be implemented using the [&lt;datalist&gt;](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist) HTML element, but this has limited functionality and behaves differently across browsers.
+`useSearchAutocomplete` helps achieve add accessible autocomplete to search components and can be styled as needed.
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Thanks, this is fixed.

File: packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+  let popoverRef = useRef<DOMRefValue<HTMLDivElement>>();
+  let unwrappedPopoverRef = useUnwrapDOMRef(popoverRef);
+  let listBoxRef = useRef();
+  let inputRef = useRef<HTMLInputElement | HTMLTextAreaElement>();
+  let domRef = useFocusableRef(ref, inputRef);
+
+  let {contains} = useFilter({sensitivity: 'base'});
+  let state = useComboBoxState(
+    {
+      ...props,
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx:74]: Good catch, fixed to undefined.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+* Virtual focus management for list box option navigation
+* Hides elements outside the input and list box from assistive technology while the list box is open in a portal
+* Custom localized announcements for option focusing, filtering, and selection using an ARIA live region to work around VoiceOver bugs
+
+## Anatomy
+
+<img src={anatomy} alt="Anatomy of useSearchAutocomplete" height={400} style={{display: 'block', margin: '20px auto'}} />
+
+A search autocomplete consists of a label, an input which displays the current value, and a list box popup. Users can type within the input
+to and see search suggestions within the list box. The list box popup may be opened by a variety of input field interactions specified
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Just fixed, thanks.

File: packages/@react-spectrum/searchfield/intl/ar-AE.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+  "clear": "مسح",
+  "invalid": "(غير صالح)",
+  "loading": "جار التحميل...",
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/intl/ar-AE.json]: Updated, thanks for finding.

File: packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+  );
+}
+
+export const Static = (props) => <Default {...props} />;
+Static.storyName = 'static items';
+
+export const DynamicItems = (props) => <Dynamic {...props} />;
+DynamicItems.storyName = 'dynamic items';
+
+export const NoItems = (props) =>  <Dynamic {...props} defaultItems={options} />;
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx]: Good find, just fixed.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+```
+
+### Dynamic collections
+
+SearchAutocomplete follows the [Collection Components API](../react-stately/collections.html), accepting both static and dynamic collections.
+The examples above show static collections, which can be used when the full list of options is known ahead of time. Dynamic collections,
+as shown below, can be used when the options come from an external data source such as an API call, or update over time.
+
+As seen below, an iterable list of options is passed to the SearchAutocomplete using the `defaultItems` prop. Each item accepts a `key` prop, which
+is passed to the `onSelectionChange` handler to identify the selected item. Alternatively, if the item objects contain an `id` property,
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Updated this, as well as fully controlled example, thanks!

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+  );
+
+  return (
+    <SearchAutocomplete
+      label="To:"
+      items={filteredItems}
+      inputValue={filterValue}
+      onInputChange={setFilterValue}
+      placeholder="Enter recipient email"
+      allowsCustomValue>
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Good find, this has been fixed.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+      {(item) => <Item>{item.email}</Item>}
+    </SearchAutocomplete>
+  );
+}
+```
+
+### Fully controlled
+
+The following example shows how you would create a controlled SearchAutocomplete, controlling everything from the selected value (`selectedKey`)
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx]: Good find, fixed this example to use onSubmit.

File: packages/@react-aria/searchfield/src/useSearchAutocomplete.ts [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        onSubmit(value, null);
+      }
+    } 
+  }, {
+    value: state.inputValue,
+    setValue: state.setInputValue
+  }, inputRef);
+  
+
+  let {listBoxProps, labelProps, inputProps: comboBoxInputProps} = useComboBox(
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/src/useSearchAutocomplete.ts:78]: Agreed, we should probably do a `useAutocomplete` hook, and have `useCombox` uses it with a text field + trigger button, and `useSearchAutocomplete` uses it with a search field. Would also need to handle the fake button for mobile. Not sure exactly where would want this hook to live though.

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 18]
```diff
@@ -0,0 +1,484 @@
...
+function Example() {
+
+  return (
+    <SearchAutocomplete label="Search with Autocomplete">
+      <Item>Aardvark</Item>
+      <Item>Kangaroo</Item>
+      <Item>Snake</Item>
+    </SearchAutocomplete>
+  );
+}
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx]: Good catch, just fixed.

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 19]
```diff
@@ -0,0 +1,484 @@
...
+  <Item>Aardvark</Item>
+  <Item>Kangaroo</Item>
+  <Item>Snake</Item>
+</SearchAutocomplete>
+```
+
+### Read only
+
+```tsx example
+<SearchAutocomplete label="Favorite Animal" isReadOnly selectedKey="red panda">
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx]: Good catch, just fixed this.

File: packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+import {Field} from '@react-spectrum/label';
+import {FocusRing} from '@react-aria/focus';
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import {ListBoxBase, useListBoxLayout} from '@react-spectrum/listbox';
+import Magnifier from '@spectrum-icons/ui/Magnifier';
+import {MobileSearchAutocomplete} from './MobileSearchAutocomplete';
+import {Placement} from '@react-types/overlays';
+import {Popover} from '@react-spectrum/overlays';
+import {ProgressCircle} from '../../progress/src/ProgressCircle';
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx]: Good catch, just fixed this.

File: packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    targetRef: inputRef,
+    overlayRef: unwrappedPopoverRef,
+    scrollRef: listBoxRef,
+    placement: `${direction} end` as Placement,
+    shouldFlip: shouldFlip,
+    isOpen: state.isOpen,
+    onClose: state.close
+  });
+
+  // Measure the width of the inputfield and the button to inform the width of the menu (below).
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx]: Thanks, fixed

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 22]
```diff
@@ -370,17 +368,25 @@ function Example() {
...
+  };
 
   return (
     <>
       <SearchAutocomplete
         label="Search engineering majors"
         defaultItems={options}
-        onSelectionChange={setMajorId}>
+        onSubmit={onSubmit}>
         {(item) => <Item>{item.name}</Item>}
```
Comment (LFDanLu) [inline packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:387]: Think you need pass the key to the Item so that your onSubmit can dig up the right major to set. Right now nothing is set when you hit enter on a major in the list

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 23]
```diff
@@ -466,12 +467,11 @@ function ControlledSearchAutocomplete() {
 
   // Specify how each of the SearchAutocomplete values should change when an
   // option is selected from the list box
-  let onSelectionChange = (key) => {
+  let onSubmit = (value, key) => {
```
Comment (LFDanLu) [inline packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:470]: hitting enter on the controlled search autocomplete after making a selection clears the field, not sure why

**Review by LFDanLu — COMMENTED** [review]
Just a couple of things I noticed after the new commits

File: packages/@react-spectrum/searchfield/chromatic/SearchAutocomplete.chromatic.tsx [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  </Grid>
+);
+
+export const PropDefaults = Template.bind({});
+PropDefaults.storyName = 'default';
+PropDefaults.args = {};
+
+export const PropSelectedKey = Template.bind({});
+PropSelectedKey.storyName = 'selectedKey: 2';
+PropSelectedKey.args = {selectedKey: '2'};
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/chromatic/SearchAutocomplete.chromatic.tsx]: Good find, removed.

File: packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx (hunk 4 above)
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/MobileSearchAutocomplete.tsx]: Just added, thanks!

File: packages/@react-spectrum/searchfield/package.json [hunk 25]
```diff
@@ -31,15 +31,39 @@
...
+    "@react-stately/combobox": "^3.0.0",
+    "@react-stately/searchfield": "^3.1.2",
+    "@react-types/button": "^3.4.0",
+    "@react-types/combobox": "^3.0.1",
+    "@react-types/overlays": "^3.5.0",
+    "@react-types/searchfield": "^3.1.1",
+    "@react-types/shared": "^3.7.1",
+    "@react-types/textfield": "^3.2.2",
+    "@spectrum-icons/ui": "^3.2.0",
+    "react-aria": "^3.8.0"
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-spectrum/searchfield/package.json:66]: I think this won't work. Will cause circular dependency issues when publishing.

[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/package.json:66]: Not sure why this happened, looking into it now...

[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/package.json]: Should be fixed now.

File: packages/@react-aria/searchfield/package.json [hunk 26]
```diff
@@ -18,14 +18,30 @@
...
-    "@react-aria/textfield": "^3.3.1",
-    "@react-aria/utils": "^3.8.2",
-    "@react-stately/searchfield": "^3.1.3",
-    "@react-types/button": "^3.4.1",
-    "@react-types/searchfield": "^3.1.2",
-    "@react-types/shared": "^3.8.0"
+    "@react-aria/combobox": "^3.0.1",
+    "@react-aria/i18n": "^3.3.1",
+    "@react-aria/interactions": "^3.4.0",
+    "@react-aria/label": "^3.1.3",
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/package.json]: Thanks, fixed along with other dependencies.

File: packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx [hunk 27]
```diff
@@ -0,0 +1,331 @@
...
+      ...props,
+      defaultFilter: contains,
+      allowsEmptyCollection: isAsync,
+      allowsCustomValue: true,
+      onSelectionChange: (key) => key !== null && props.onSubmit && props.onSubmit(null, key)
+    }
+  );
+  let layout = useListBoxLayout(state);
+  
+  let {inputProps, listBoxProps, labelProps, clearButtonProps} = useSearchAutocomplete(
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/src/SearchAutocomplete.tsx:86]: Good idea, added to follow-up (#2286)

**Comment by devongovett** [comment]
Danni and I were discussing, and we think it would be good to move this into its own package for now (`@react-aria/autocomplete`, and `@react-spectrum/autocomplete`). We aren't sure exactly when we'll have time to finish this off, and would like to avoid putting the search field packages back into alpha (which will show up in the docs) for the time being. It may end up moving again, but for now I think a separate package would be better. Sorry to change course! Hopefully not too much work to move it. 😄

File: packages/@react-types/searchfield/package.json [hunk 28]
```diff
@@ -9,7 +9,9 @@
     "url": "https://github.com/adobe/react-spectrum"
   },
   "dependencies": {
-    "@react-types/textfield": "^3.2.3"
+    "@react-types/combobox": "^3.0.1",
+    "@react-types/shared": "^3.7.1",
+    "@react-types/textfield": "^3.2.1"
```
Comment (snowystinger) [inline packages/@react-types/searchfield/package.json:14]: not sure this should be less than it used to be, what caused this change?

**Comment by devongovett** [comment]
Superseded by #2371.

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 29]
```diff
@@ -0,0 +1,600 @@
...
+  the first or last item accordingly
+* Support for opening the list box when typing, on focus, or manually
+* Handles virtual clicks on the input from touch screen readers to toggle the list box
+* Virtual focus management for list box option navigation
+* Hides elements outside the input and list box from assistive technology while the list box is open in a portal
+* Custom localized announcements for option focusing, filtering, and selection using an ARIA live region to work around VoiceOver bugs
+
+## Anatomy
+
+<img src={anatomy} alt="Anatomy of useSearchAutocomplete" height={400} style={{display: 'block', margin: '20px auto'}} />
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:72]: Added to follow-up (#2286)

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 30]
```diff
@@ -0,0 +1,600 @@
...
+* Support for disabled options
+* Support for groups of items in sections
+* Support for custom user input values
+* Support for controlled and uncontrolled options, selection, input value, and open state
+* Support for custom filter functions
+* Async loading and infinite scrolling support
+* Support for virtualized scrolling for performance with long lists
+* Exposed to assistive technology as a `combobox` with ARIA
+* Labeling support for accessibility
+* Required and invalid states exposed to assistive technology via ARIA
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:60]: Added investigation/discussion to follow-up (#2286)

File: packages/@react-types/searchfield/src/index.d.ts [hunk 31]
```diff
@@ -20,5 +23,49 @@ export interface SearchFieldProps extends TextFieldProps {
...
+  /** The list of SearchAutocomplete items (controlled). */
+  items?: Iterable<T>,
+  /** Method that is called when the open state of the menu changes. Returns the new open state and the action that caused the opening of the menu. */
+  onOpenChange?: (isOpen: boolean, menuTrigger?: MenuTriggerAction) => void,
+  /** The value of the SearchAutocomplete input (controlled). */
+  inputValue?: string,
+  /** The default value of the SearchAutocomplete input (uncontrolled). */
+  defaultInputValue?: string,
+  /** Handler that is called when the SearchAutocomplete input value changes. */
+  onInputChange?: (value: string) => void,
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-types/searchfield/src/index.d.ts:38]: Added to follow-up (#2286)

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx [hunk 32]
```diff
@@ -0,0 +1,600 @@
...
+}
+
+<SearchAutocomplete label="Search Animals">
+  <Item key="red panda">Red Panda</Item>
+  <Item key="cat">Cat</Item>
+  <Item key="dog">Dog</Item>
+  <Item key="aardvark">Aardvark</Item>
+  <Item key="kangaroo">Kangaroo</Item>
+  <Item key="snake">Snake</Item>
+</SearchAutocomplete>
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:196]: Added investigation to follow-up (#2286)

File: packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx [hunk 33]
```diff
@@ -0,0 +1,188 @@
...
+  onInputChange: action('onInputChange'),
+  onBlur: action('onBlur'),
+  onFocus: action('onFocus'),
+  onChange: action('onChange'),
+  onSubmit: action('onSubmit')
+};
+
+function Default(props) {
+  return (
+    <SearchAutocomplete label="Search with Autocomplete" {...mergeProps(props, actions)}>
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx:55]: Added to follow-up (#2286)

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx (hunk 6 above)
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:118]: Added to follow-up (#2286)

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 34]
```diff
@@ -0,0 +1,484 @@
...
+<SearchAutocomplete label="Favorite Animal" menuTrigger="manual">
+  <Item>Red Panda</Item>
+  <Item>Cat</Item>
+  <Item>Dog</Item>
+  <Item>Aardvark</Item>
+  <Item>Kangaroo</Item>
+  <Item>Snake</Item>
+</SearchAutocomplete>
+```
+
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:393]: Added to follow-up (#2286)

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 35]
```diff
@@ -0,0 +1,484 @@
...
+      loadingState={list.loadingState}
+      onLoadMore={list.loadMore}>
+      {item => <Item key={item.name}>{item.name}</Item>}
+    </SearchAutocomplete>
+  );
+}
+```
+
+## Validation
+SearchAutocomplete can display a validation state to communicate to the user whether the current value is valid or invalid.
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:279]: Added to follow-up (#2286)

File: packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx [hunk 36]
```diff
@@ -0,0 +1,484 @@
...
+  <Item>Aardvark</Item>
+  <Item>Kangaroo</Item>
+  <Item>Snake</Item>
+</SearchAutocomplete>
+```
+
+### Custom widths
+
+```tsx example
+<SearchAutocomplete label="Favorite Animal" width="size-6000" maxWidth="100%">
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/docs/SearchAutocomplete.mdx:459]: Added to follow-up (#2286)

File: packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx [hunk 37]
```diff
@@ -0,0 +1,188 @@
...
+directionTop.storyName = 'direction: top';
+
+export const customWidth500 = (props) => <Default {...props} width="size-500" />;
+customWidth500.storyName = 'custom width: size-500';
+
+export const customWidth3000 = (props) => <Default {...props} width="size-3000'" />;
+customWidth3000.storyName = 'custom width: size-3000';
+
+export const customWidth6000 = (props) => <Default {...props} width="size-6000" />;
+customWidth6000.storyName = 'custom width: size-6000';
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-spectrum/searchfield/stories/SearchAutocomplete.stories.tsx:185]: Added to follow-up (#2286)

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx (hunk 22 above)
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:387]: Added to follow-up (#2286)

File: packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx (hunk 23 above)
Comment (reidbarber) [reply packages/@react-aria/searchfield/docs/useSearchAutocomplete.mdx:470]: Added to follow-up (#2286)

File: packages/@react-types/searchfield/package.json (hunk 28 above)
Comment (reidbarber) [reply packages/@react-types/searchfield/package.json:14]: Handled by creating autocomplete package via #2371.

File: packages/@react-aria/searchfield/src/useSearchAutocomplete.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/@react-aria/searchfield/src/useSearchAutocomplete.ts:78]: Will be handled in the future with autocomplete package added via #2371.
